media/client/main/source/WebAudioPlayer.cpp

```diff
@@ -60,83 +67,163 @@ std::unique_ptr<IWebAudioPlayer> WebAudioPlayerFactory::createWebAudioPlayer(std
 namespace firebolt::rialto::client
 {
 WebAudioPlayer::WebAudioPlayer(std::weak_ptr<IWebAudioPlayerClient> client, const std::string &audioMimeType,
-                               const uint32_t priority, const WebAudioConfig *config)
+                               const uint32_t priority, const WebAudioConfig *config,
+                               const std::shared_ptr<IWebAudioPlayerIpcFactory> &webAudioPlayerIpcFactory,
+                               const std::shared_ptr<ISharedMemoryManagerFactory> &sharedMemoryManagerFactory)
     : m_webAudioPlayerClient(client)
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
+    m_sharedMemoryManager = sharedMemoryManagerFactory->getSharedMemoryManager();
+    if (!m_sharedMemoryManager)
+    {
+        throw std::runtime_error("Shared memory manager could not be created");
+    }
+
+    if (!m_sharedMemoryManager->registerClient(this))
+    {
+        throw std::runtime_error("Failed to register client with SharedMemoryManager");
+    }
+
+    if (!webAudioPlayerIpcFactory)
+    {
+        throw std::runtime_error("Web audio player ipc factory could not be null");
+    }
+
+    m_webAudioPlayerIpc = webAudioPlayerIpcFactory->createWebAudioPlayerIpc(this, audioMimeType, priority, config);
+    if (!m_webAudioPlayerIpc)
+    {
+        (void)m_sharedMemoryManager->unregisterClient(this);
+        throw std::runtime_error("Web audio player ipc could not be created");
+    }
 }
 
 WebAudioPlayer::~WebAudioPlayer()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
+
+    m_webAudioPlayerIpc.reset();
+    if (!m_sharedMemoryManager->unregisterClient(this))
+    {
+        RIALTO_CLIENT_LOG_WARN("Failed to unregister client with SharedMemoryManager");
+    }
+
+    m_sharedMemoryManager.reset();
 }
 
 bool WebAudioPlayer::play()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->play();
 }
 
 bool WebAudioPlayer::pause()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->pause();
 }
 
 bool WebAudioPlayer::setEos()
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
 
-    return false;
+    return m_webAudioPlayerIpc->setEos();
 }
 
-bool WebAudioPlayer::getBufferAvailable(uint32_t &availableFrames, std::shared_ptr<WebAudioShmInfo> &webAudioShmInfo)
+bool WebAudioPlayer::getBufferAvailable(uint32_t &availableFrames, std::shared_ptr<WebAudioShmInfo> &)
 {
     RIALTO_CLIENT_LOG_DEBUG("entry:");
-
-    return false;
+    if (!m_webAudioShmInfo)
+    {
+        m_webAudioShmInfo = std::make_shared<WebAudioShmInfo>();
+    }
+    std::lock_guard<std::mutex> bufLocker(m_bufLock);
```

Comment:
WebAudioPlayerIpc::getBufferAvailable just sends the message to rialto server and waits for response. Do we need a lock here?

Reply:
This lock is used to syncronise writeBuffer and getBufferAvailable. We cannot have both in progress at the same time as they both use m_webAudioShmInfo.

Reply:
The mutex prevents updating  m_webAudioShmInfo and at the same time using it in writeBuffer in case the client calls    getBufferAvailable/writeBuffer asynchronously